File: nim/grade-school/grade_school.nim
import algorithm, sequtils

type
  Student* = object
    name*: string
    grade*: int

  School* = object
    students*: seq[Student]

proc showGrades(school: School): seq[int] =
  result = school.students.mapIt(it.grade).deduplicate
  result.sort

proc grade*(school: School, grade: int): seq[string] =
  ## Returns a sorted sequence of students names, filtered by the
  ## given `grade` from the given `school` 
  result = school.students.filterIt(it.grade == grade).mapIt(it.name)
  result.sort

proc roster*(school: School): seq[string] =
  ## Returns a sequence of student names in the `school`
  for g in school.showGrades:
    result.add(school.grade(g))

proc addStudent*(school: var School, name: string, grade: int) =
  ## Adds a `Student` to the given `school` using the given `name` and `grade`
  ## 
  ## Raises a `ValueError` if the `Student` already exists in the `school` 
  if name notin school.roster:
    school.students.add(Student(name: name, grade: grade))
  else:
    raise newException(ValueError, "This student already exists in the roster.")

File: ocaml/grade-school/grade_school.ml
open Base
module Int_map = Map.M (Int)

type school = string list Int_map.t

let empty_school = Map.empty (module Int)

let add name grade = Map.add_multi ~key:grade ~data:name

let grade = Fn.flip Map.find_multi

let sorted = Map.map ~f:(List.sort ~compare:String.compare)

let roster school = sorted school |> Map.data |> List.concat
